includes/admin/class-wc-rest-payments-files-controller.php

```diff
@@ -61,20 +60,33 @@ public function upload_file( $request ) {
 	 * @return WP_Error|WP_HTTP_Response
 	 */
 	public function get_file( WP_REST_Request $request ) {
-		$file_id = $request->get_param( 'file_id' );
-		$result  = $this->forward_request( 'get_file_contents', [ $file_id ] );
+		$file_id      = $request->get_param( 'file_id' );
+		$file_service = new WC_Payments_File_Service();
+		$purpose      = get_transient( WC_Payments_File_Service::CACHE_KEY_PREFIX_PURPOSE . $file_id );
 
-		if ( is_wp_error( $result ) ) {
-			$error_status_code = 'resource_missing' === $result->get_error_code() ? WP_Http::NOT_FOUND : WP_Http::INTERNAL_SERVER_ERROR;
-			return rest_ensure_response(
-				new WP_Error(
-					$result->get_error_code(),
-					$result->get_error_message(),
-					[ 'status' => $error_status_code ]
-				)
+		if ( ! $purpose ) {
+			$file = $this->forward_request( 'get_file', [ $file_id ] );
+			if ( is_wp_error( $file ) ) {
+				return $this->file_error_response( $file );
+			}
+			$purpose = $file->data['purpose'];
+			set_transient( WC_Payments_File_Service::CACHE_KEY_PREFIX_PURPOSE . $file_id, $purpose, WC_Payments_File_Service::CACHE_PERIOD );
+		}
+
+		if ( $file_service->file_need_access_permissions( $purpose ) && ! $this->check_permission() ) {
```

Comment:
Shouldn't it be an OR condition?

Reply:
no, we need both conditions. `file_need_access_permissions` checks if the file shouldn't be public, and `check_permission` if the user has required permissions. So, if the file isn't public **and** the user does not have required permissions, we should throw access denied response.

Reply:
I see. Thanks for clarifying. In that case, `check_permission` is named oddly. I thought it meant a user _has_ permission if it returned true.